src/random-seed/random-seed.c

```diff
@@ -306,6 +308,69 @@ static int save_seed_file(
         return 0;
 }
 
+static int refresh_boot_seed(void) {
+        uint8_t buffer[RANDOM_EFI_SEED_SIZE];
+        struct sha256_ctx hash_state;
+        _cleanup_free_ char *esp_path = NULL, *seed_path = NULL;
+        _cleanup_free_ void *seed_file_bytes = NULL;
+        _cleanup_close_ int seed_fd = -1;
+        size_t len;
+        ssize_t r;
+
+        assert_cc(RANDOM_EFI_SEED_SIZE == SHA256_DIGEST_SIZE);
+
+        /* We're doing this opportunistically, so if the seeding dance before didn't manage to initialize the
+         * RNG, there's no point in doing it here. Secondly, getrandom(GRND_NONBLOCK) has been around longer
+         * than EFI seeding anyway, so there's no point in having non-getrandom() fallbacks here. So if this
+         * fails, just return early to cut our losses. */
+        r = getrandom(buffer, sizeof(buffer), GRND_NONBLOCK);
+        if (r != sizeof(buffer))
+                return r < 0 ? -errno : -EIO;
+        sha256_init_ctx(&hash_state);
+        sha256_process_bytes(&r, sizeof(r), &hash_state);
+        sha256_process_bytes(buffer, r, &hash_state);
+
+        /* We search for ESP with a NULL path and root, and in unprivileged mode, so that we don't emit any
+         * warnings if this fails. It's just opportunistic, so it doesn't matter in that case. */
+        r = find_esp_and_warn(NULL, NULL, true, &esp_path, 0, 0, 0, NULL, NULL);
+        if (r)
+                return r;
+        seed_path = path_join(esp_path, "/loader/random-seed");
+        if (!seed_path)
+                return log_oom();
+
+        /* Read the existing seed, but if it's not there, return early, since it means bootctl hasn't yet set
+         * up random seeds for the loader. */
+        seed_fd = open(seed_path, O_RDWR|O_CLOEXEC|O_NOCTTY);
+        if (seed_fd < 0)
+                return -errno;
```

Comment:
maybe use:

```c
seed_fd = chase_symlinks_and_open("/loader/random-seed", esp_path, CHASE_PREFIX_ROOT, O_RDWR|O_CLOEXEC|O_NOCTTY);
```

this is an extra safety precaution in case the ESP is not a VFAT, but some fs that allows symlinks (which you can do, by loading an efifs driver into the firmware). In that case we really want to make sure the symlinks cannot be used to leave the ESP, i.e. use this to overwrite random stuff by creating the right stuff in the ESP.

(after all, let's not forget: the ESP is untrusted territory in SecureBoot environments)

(bootctl doesn't strictly use chase_symlinks everywhere right now, but it really should, to deal with such attacks)

Reply:
Neat. Will do.